### PR TITLE
58 implement parse and or

### LINKED ISSUE #58: Create `parse_and_or`

### PR DESCRIPTION
**Pull request by rask24**
fix #58

- **Implement parse_list**
- **Add tests for parse_list**
- **Fix parse_pipeline**
- **Add complex test to test_parse_list**

## Release Notes

- **New Features**
	- Introduced parsing capabilities for logical operations in command sequences.
	- Added support for new grammar rules allowing complex command structures with logical AND (`&&`) and OR (`||`) operators.
	- Expanded abstract syntax tree (AST) functionality with a new node type for improved error handling.

- **Bug Fixes**
	- Enhanced parsing logic to prevent errors with unexpected tokens, improving robustness.

- **Tests**
	- Implemented unit tests for the new parsing function, validating various command structures and ensuring accurate AST representation.

### CODE REVIEW

**Comment by alex216** [comment]
suggestion
- change `parse_pipeline()` -> `parse_list()`
https://github.com/rask24/minishell/blob/39b88a71aa8e38b4db26b1e6cc3be1b89e5ce387/src/parser/parser.c#L23
- apply `return_entire_path()` to `argv[0]` 
https://github.com/rask24/minishell/blob/39b88a71aa8e38b4db26b1e6cc3be1b89e5ce387/src/exec/execute_simple_command.c#L83

File: tests/unit/test_parse_list.cpp [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+  ft_lstadd_back(&token_list, construct_token(TOKEN_EOF, nullptr));
+
+  t_ast *node = parse_list(&token_list);
+
+  EXPECT_EQ(node, nullptr);
+
+  destroy_token_list(token_list);
+}
+
+TEST(parse_list, InvalidTOken2) {
```
Comment (alex216) [inline tests/unit/test_parse_list.cpp:245]: typo `InvalidTOken2` -> `InvalidToken2`

**Comment by rask24** [comment]
> suggestion
> 
> * ~change `parse_pipeline()` -> `parse_list()`~
>   ~https://github.com/rask24/minishell/blob/39b88a71aa8e38b4db26b1e6cc3be1b89e5ce387/src/parser/parser.c#L23
>   ~
> * apply `return_entire_path()` to `argv[0]`
>   https://github.com/rask24/minishell/blob/39b88a71aa8e38b4db26b1e6cc3be1b89e5ce387/src/exec/execute_simple_command.c#L83

I'll handle this in #63.

**Comment by alex216** [comment]
- suggestion
  - refactor 3 parse files for simplify error handling.
- motivation
  - handle all syntax errors inside `parse_simple_command.c` as possible.
  https://github.com/rask24/minishell/blob/d3a9a4a979e16ab70c970bb047171557afba62d9/src/parser/parse_list.c#L55

- plan1
```.c
t_ast	*parse_list(t_token_list **cur_token)
{
	t_ast			*node;
	t_ast 			*tmp;
	t_ast_node_type	node_type;

	node = parse_pipeline(cur_token);
	if (node == NULL)
		return (NULL);
	while (get_token_type(*cur_token) != TOKEN_EOF)
	{
		node_type = convert_token_to_node(get_token_type(*cur_token));
		if (node_type != AST_AND && node_type != AST_OR)
			return (handle_error(node, get_token_value(*cur_token)));
		consume_token(cur_token);
		tmp = parse_pipeline(cur_token); // preserve same structure with parse_pipeline()
		if (!tmp)
			return (handle_error(node, get_token_value(*cur_token)));
		node = construct_ast(node_type, node, tmp);
	}
	return (node);
}
```
```.c

// does not need to separately check syntax error in parse_pipeline.c file
// parse_pipeline.c only need to recognize pipeline, AND_IF, AND_OR, and nothing more.

// static bool	is_pipeline_head_token(t_token_type type)
// {
// 	return (type == TOKEN_WORD
// 		|| type == TOKEN_LESS
// 		|| type == TOKEN_GREAT
// 		|| type == TOKEN_DLESS
// 		|| type == TOKEN_DGREAT);
// }

t_ast	*parse_pipeline(t_token_list **cur_token)
{
	t_ast	*node;
	t_ast	*tmp;

	node = parse_simple_command(cur_token);
	if (node == NULL)
		return (NULL);
	while (get_token_type(*cur_token) != TOKEN_EOF)
	{
		if (get_token_type(*cur_token) != TOKEN_PIPE)
			break ;
		consume_token(cur_token); // error check moved to parse_simple_command()
		tmp = parse_simple_command(cur_token); // preserve same structure with parse_list()
		if (!tmp)
			return (handle_error(node, get_token_value(*cur_token)));
		node = construct_ast(AST_PIPE, node, tmp);
	}
	return (node);
}
```
```.c
t_ast	*parse_simple_command(t_token_list **cur_token)
{
	t_ast	*node;
	bool	is_valid;

        // check all syntax error here
	if (!(is_redirect_token(get_token_type(*cur_token)) || get_token_type(*cur_token) == TOKEN_WORD))
		return (handle_error(NULL, get_token_value(*cur_token)));
	node = construct_ast(AST_COMMAND, NULL, NULL);
	is_valid = true;
	while (get_token_type(*cur_token) != TOKEN_EOF)
	{
		if (get_token_type(*cur_token) == TOKEN_WORD)
			is_valid = parse_word(node, cur_token);
		else if (is_redirect_token(get_token_type(*cur_token)))
			is_valid = parse_redirect(node, cur_token);
		else
			break ;
		if (!is_valid)
			return (handle_error(node, get_token_value(*cur_token)));
	}
	return (node);
}

```
- plan2
  - handle these concerns after all the syntax is defined.

**Review by alex216 — APPROVED** [review]
@rask24 ok!

**Pull request merged**